### PR TITLE
[Compatibility] Accept options argument to `Regexp.{new,compile}` of String; Warn for unknown types

### PR DESCRIPTION
**Pull request by rwstauner**
Regexp.new now supports passing the regexp flags not only as an Integer,
but also as a String. Unknown flags raise ArgumentError.
Otherwise, anything other than true, false, nil or Integer will be warned.
[[Feature #18788](https://bugs.ruby-lang.org/issues/18788)]

### CODE REVIEW

**Comment by andrykonchin** [comment]
CI failures seem irrelevant to the changes.

**Comment by eregon** [comment]
> CI failures seem irrelevant to the changes.

Right, it seems related to a new ECJ in latest mx. And the workflow here just uses latest mx instead of the one specified in `common.json`. Maybe we should try to use the correct mx in GitHub Actions too.
But anyway we need to fix these ECJ warnings, I'll do it.

File: src/main/ruby/truffleruby/core/regexp.rb [hunk 1]
```diff
@@ -144,8 +144,30 @@ def self.new(pattern, opts = undefined, encoding = nil)
...
+      end
+    elsif !Primitive.undefined?(opts)
+      # Valid values are true, false, nil.
+      # Other values warn but treat as true.
+      if Primitive.false?(opts) || Primitive.nil?(opts)
+        opts = 0
+      else
+        if !Primitive.true?(opts)
+          warn "expected true or false as ignorecase: #{opts}"
+        end
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/regexp.rb:168]: This seems like a good place to use suffix `unless`: `warn unless Primitive.true?(opts)`

**Review by eregon — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +19,7 @@ Bug fixes:
 * Fix using the `--cpusampler` profiler when there are custom unblock functions for `rb_thread_call_without_gvl()` (#3013, @eregon).
 * Fix recursive raising `FrozenError` exception when redefined `#inspect` modifies an object (#3388, @andrykonchin).
 * Fix `Integer#div` returning the wrong object type when the divisor is a `Rational` (@simonlevasseur, @nirvdrum).
+* Accept options argument to `Regexp.{new,compile}` of String; Warn for unknown types. (@rwstauner).
```
[the remark this answers is not in the document]
Comment (rwstauner) [reply CHANGELOG.md:22]: Oh, this the id of the github issue where it was logged? 👍

[the remark this answers is not in the document]
Comment (eregon) [reply CHANGELOG.md:22]: Yep: https://github.com/oracle/truffleruby/issues/3039

File: src/main/ruby/truffleruby/core/regexp.rb [hunk 3]
```diff
@@ -144,8 +144,30 @@ def self.new(pattern, opts = undefined, encoding = nil)
...
+      opts = opts.chars.reduce(0) do |result, opt|
+        case opt
+        when 'i'
+          result | IGNORECASE
+        when 'm'
+          result | MULTILINE
+        when 'x'
+          result | EXTENDED
+        else
+          raise ArgumentError, "unknown regexp option: #{opt}"
```
[the remark this answers is not in the document]
Comment (rwstauner) [reply src/main/ruby/truffleruby/core/regexp.rb:157]: Updated the spec to test this

File: CHANGELOG.md [hunk 4]
```diff
@@ -35,6 +35,7 @@ Compatibility:
 * Implement `MatchData#{byteoffset,deconstruct,deconstruct_keys}` from Ruby 3.2 (#3039, @rwstauner).
 * Add `Integer#ceildiv` method (#3039, @simonlevasseur, @nirvdrum).
 * Implement `Class#attached_object` method (#3039, @andrykonchin).
+* Accept options argument to `Regexp.{new,compile}` of String; Warn for unknown types. (#3039, @rwstauner).
```
Comment (andrykonchin) [inline CHANGELOG.md:38]:
```suggestion
* Accept options argument to `Regexp.{new,compile}` of String; Warn for unknown types (#3039, @rwstauner).
```

**Review by andrykonchin — APPROVED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (andrykonchin) [reply CHANGELOG.md:38]: Will change it myself.

**Pull request merged**